snowfall/training/asg_graph.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (c)  2020  Xiaomi Corp.       (author: Fangjun Kuang)
+
+from functools import lru_cache
+from typing import Iterable
+from typing import List
+
+import re
+
+import k2
+import torch
+
+from .ctc_graph import build_ctc_topo
+
+
+def filter_disambig_symbols(symbol_table: k2.SymbolTable,
+                            pattern: str = '#') -> List[int]:
+    '''Return a list of phone IDs containing no disambiguation symbols.
+
+    Caution:
+      You may need to remove the ID with value 0 if the return value
+      is used to create a phone LM.
+
+    Args:
+      symbol_table:
+        A symbol table in k2.
+      pattern:
+        Symbols containing this pattern are disambiguation symbols.
+    Returns:
+      Return a list of symbol IDs excluding those from disambiguation symbols.
+    '''
+    regex = re.compile(pattern)
+    symbols = symbol_table.symbols
+    ans = []
+    for s in symbols:
+        if not regex.search(s):
+            ans.append(symbol_table[s])
+    return ans
+
+
+def create_bigram_phone_lm(phones: List[int]) -> k2.Fsa:
+    '''Create a bigram phone LM.
+    The resulting FSA (P) has a start-state and a state for
+    each phone 0, 1, ....; and each of the above-mentioned states
+    has a transition to the state for each phone and also to the final-state.
+
+    Caution:
+      blank is not a phone.
+    '''
+    final_state = len(phones) + 1
+    rules = ''
+    for i in range(1, final_state):
+        rules += f'0 {i} {phones[i-1]} 0.0\n'
+
+    for i in range(1, final_state):
+        for j in range(1, final_state):
+            rules += f'{i} {j} {phones[j-1]} 0.0\n'
+        rules += f'{i} {final_state} -1 0.0\n'
+    rules += f'{final_state}'
+    print(len(rules.split('\n')))
+    return k2.Fsa.from_str(rules)
+
+
+class AsgTrainingGraphCompiler(object):
+
+    def __init__(self,
+                 L_inv: k2.Fsa,
+                 phones: k2.SymbolTable,
```

Comment:
I think it might make sense to have a more "official", future-proof way to label the elements of a symbol table as disambig symbols, so that we could decide in the data-prep.   At a minimum the default pattern should be stronger, e.g. '^#\d+$', but IDK if that's a valid pattern in python `re`.

Reply:
I think so if it is a raw string. Let me try it.